File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiSuggestionMenu.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Input;
using Microsoft.UI.Xaml.Media;
using Maple.UI.Atoms;
using Windows.System;

namespace Maple.UI
{
    /// <summary>One row in a Suggestion Menu (e.g. an @-mention candidate).</summary>
    public sealed record MuiSuggestionItem(string Id, string Label, string? IconName = null, string? Meta = null);

    /// <summary>
    /// Maple.UI Suggestion Menu molecule (unified-component-catalog.md
    /// §2.4, "Suggestion Menu" row: "Query-driven autocomplete list", built
    /// from Popover, Icon, Text) — composes <see cref="MuiPopover"/>; the
    /// caller already filters <see cref="Items"/> by the live query text
    /// (this owns only the anchored presentation and keyboard/mouse
    /// selection, matching `mui-suggestion-menu.component.ts`'s own split).
    ///
    /// Every (re)open or re-filter highlights row 0 — "Enter picks the top
    /// match" per the usual @-mention/autocomplete convention — without the
    /// caller having to manage that state itself.
    /// </summary>
    public sealed class MuiSuggestionMenu : ContentControl
    {
        public static readonly DependencyProperty IsOpenProperty =
            DependencyProperty.Register(nameof(IsOpen), typeof(bool), typeof(MuiSuggestionMenu),
                new PropertyMetadata(false, (d, _) => ((MuiSuggestionMenu)d).OnIsOpenChanged()));

        public static readonly DependencyProperty PlacementProperty =
            DependencyProperty.Register(nameof(Placement), typeof(MuiPopoverPlacement), typeof(MuiSuggestionMenu),
                new PropertyMetadata(MuiPopoverPlacement.Bottom, (d, e) => ((MuiSuggestionMenu)d)._popover.Placement = (MuiPopoverPlacement)e.NewValue));

        public static readonly DependencyProperty ItemsProperty =
            DependencyProperty.Register(nameof(Items), typeof(IReadOnlyList<MuiSuggestionItem>), typeof(MuiSuggestionMenu),
                new PropertyMetadata(null, (d, _) => ((MuiSuggestionMenu)d).OnItemsChanged()));

        public bool IsOpen
        {
            get => (bool)GetValue(IsOpenProperty);
            set => SetValue(IsOpenProperty, value);
        }

        public MuiPopoverPlacement Placement
        {
            get => (MuiPopoverPlacement)GetValue(PlacementProperty);
            set => SetValue(PlacementProperty, value);
        }

        public IReadOnlyList<MuiSuggestionItem>? Items
        {
            get => (IReadOnlyList<MuiSuggestionItem>?)GetValue(ItemsProperty);
            set => SetValue(ItemsProperty, value);
        }

        public event EventHandler<string>? ItemSelected;
        public event EventHandler? CloseRequested;

        private readonly MuiPopover _popover = new();
        private readonly StackPanel _list = new() { Orientation = Orientation.Vertical, Spacing = 2, MinWidth = 200 };
        private int _activeIndex;

        public MuiSuggestionMenu()
        {
            Content = _popover;
            IsTabStop = false;
            IsHitTestVisible = false;

            _popover.PanelContent = _list;
            _popover.CloseRequested += (_, _) => CloseRequested?.Invoke(this, EventArgs.Empty);
            _list.KeyDown += OnKeyDown;

            Rebuild();
        }

        private void OnIsOpenChanged()
        {
            if (IsOpen) _activeIndex = 0;
            _popover.IsOpen = IsOpen;
            RebuildHighlights();
        }

        private void OnItemsChanged()
        {
            if (IsOpen) _activeIndex = 0;
            Rebuild();
        }

        private int Count => Items?.Count ?? 0;

        private void OnKeyDown(object sender, KeyRoutedEventArgs e)
        {
            if (Count == 0) return;
            if (e.Key == VirtualKey.Down)
            {
                e.Handled = true;
                SetActive(MuiMenuNavMath.WrapIndex(_activeIndex, 1, Count));
            }
            else if (e.Key == VirtualKey.Up)
            {
                e.Handled = true;
                SetActive(MuiMenuNavMath.WrapIndex(_activeIndex, -1, Count));
            }
            else if (e.Key == VirtualKey.Enter)
            {
                e.Handled = true;
                if (_activeIndex < Count) SelectItem(Items![_activeIndex]);
            }
        }

        private void SetActive(int index)
        {
            _activeIndex = index;
            RebuildHighlights();
        }

        private void SelectItem(MuiSuggestionItem item) => ItemSelected?.Invoke(this, item.Id);

        private static Brush R(string key) => (Brush)Application.Current.Resources[key];

        private void Rebuild()
        {
            _list.Children.Clear();
            foreach (var item in Items ?? Array.Empty<MuiSuggestionItem>())
                _list.Children.Add(BuildRow(item));
            RebuildHighlights();
        }

        private Border BuildRow(MuiSuggestionItem item)
        {
            var content = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8 };
            if (!string.IsNullOrEmpty(item.IconName))
                content.Children.Add(new MuiIcon { IconName = item.IconName, Size = MuiIconSize.Sm16 });
            content.Children.Add(new MuiText { Text = item.Label, Variant = MuiTextVariant.RowLabel });
            if (!string.IsNullOrEmpty(item.Meta))
                content.Children.Add(new MuiText { Text = item.Meta, Variant = MuiTextVariant.ValueChip });

            var row = new Border
            {
                CornerRadius = new CornerRadius(6),
                Padding = new Thickness(10, 7, 10, 7),
                Child = content,
            };
            row.PointerPressed += (_, _) => SelectItem(item);
            return row;
        }

        private void RebuildHighlights()
        {
            for (var i = 0; i < _list.Children.Count; i++)
                if (_list.Children[i] is Border row)
                    row.Background = i == _activeIndex ? R("MapleSurfaceHover") : new SolidColorBrush(Microsoft.UI.Colors.Transparent);
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/Molecules/MuiSuggestionMenu.swift
// MuiSuggestionMenu.swift — Maple UI Molecules-L1 (unified-component-
// catalog.md §2.4; Built from: Popover, Icon, Text). A query-driven
// autocomplete list (e.g. an @-mention picker) — the caller already
// filters `items` by the live query text; this component owns only the
// anchored presentation and keyboard/mouse selection.

import SwiftUI

public struct MuiSuggestionItem: Identifiable, Sendable {
    public let id: String
    public let label: String
    public let icon: String?
    public let meta: String?

    public init(id: String, label: String, icon: String? = nil, meta: String? = nil) {
        self.id = id
        self.label = label
        self.icon = icon
        self.meta = meta
    }
}

public struct MuiSuggestionMenu<Trigger: View>: View {
    @Binding public var open: Bool
    public let placement: MuiPopoverPlacement
    public let items: [MuiSuggestionItem]
    public let select: (String) -> Void
    @ViewBuilder public let trigger: Trigger

    @State private var activeIndex = 0

    public init(
        open: Binding<Bool>,
        placement: MuiPopoverPlacement = .bottom,
        items: [MuiSuggestionItem],
        select: @escaping (String) -> Void,
        @ViewBuilder trigger: () -> Trigger
    ) {
        self._open = open
        self.placement = placement
        self.items = items
        self.select = select
        self.trigger = trigger()
    }

    public var body: some View {
        trigger
            .muiPopover(isPresented: open, placement: placement, closeRequested: { open = false }) {
                menu
            }
            // A freshly (re)opened or re-filtered list always highlights
            // its first row, matching the usual @-mention/autocomplete
            // convention of "Enter picks the top match".
            .onChange(of: open) { _, isOpen in if isOpen { activeIndex = 0 } }
            .onChange(of: items.map(\.id)) { activeIndex = 0 }
    }

    private var menu: some View {
        VStack(alignment: .leading, spacing: 2) {
            if items.isEmpty {
                MuiText("No matches", variant: .body, color: .muted)
                    .padding(.horizontal, MuiTokens.spacingSm)
                    .padding(.vertical, MuiTokens.spacingXs)
            }
            ForEach(Array(items.enumerated()), id: \.element.id) { index, item in
                row(item, index: index)
            }
        }
        .frame(minWidth: 200)
        #if os(macOS)
        .focusable()
        .onKeyPress(.downArrow) { moveActive(1); return .handled }
        .onKeyPress(.upArrow) { moveActive(-1); return .handled }
        .onKeyPress(.return) { activateHighlighted(); return .handled }
        #endif
    }

    private func row(_ item: MuiSuggestionItem, index: Int) -> some View {
        Button {
            select(item.id)
            open = false
        } label: {
            HStack(spacing: MuiTokens.spacingSm) {
                if let icon = item.icon {
                    MuiIcon(name: icon, size: .sm, color: MuiTokens.textMuted)
                }
                MuiText(item.label, variant: .rowLabel)
                Spacer(minLength: MuiTokens.spacingSm)
                if let meta = item.meta {
                    MuiText(meta, variant: .body, color: .muted)
                }
            }
            .padding(.horizontal, MuiTokens.spacingSm)
            .padding(.vertical, MuiTokens.spacingXs)
            .frame(minHeight: 32)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                index == activeIndex ? MuiTokens.surfaceHover : .clear,
                in: RoundedRectangle(cornerRadius: MuiTokens.radiusSm, style: .continuous)
            )
        }
        .buttonStyle(.plain)
    }

    private func moveActive(_ direction: Int) {
        activeIndex = MuiMenuNavMath.wrappedIndex(current: activeIndex, direction: direction, count: items.count)
    }

    private func activateHighlighted() {
        guard items.indices.contains(activeIndex) else { return }
        select(items[activeIndex].id)
        open = false
    }
}

#Preview("MuiSuggestionMenu") {
    struct Demo: View {
        @State private var open = false

        var body: some View {
            MuiSuggestionMenu(
                open: $open,
                items: [
                    MuiSuggestionItem(id: "ada", label: "Ada Lovelace", icon: "person.crop.circle"),
                    MuiSuggestionItem(id: "grace", label: "Grace Hopper", icon: "person.crop.circle"),
                    MuiSuggestionItem(id: "kat", label: "Katherine Johnson", icon: "person.crop.circle"),
                ],
                select: { _ in }
            ) {
                MuiButton(label: "@mention", variant: .secondary) { open.toggle() }
            }
        }
    }
    return Demo()
        .padding(80)
        .background(MuiTokens.bg)
}
